### PR TITLE
Signed comparisons

### PR DESCRIPTION
**Pull request by 4l0n50**

### CODE REVIEW

File: prover/src/opcodes/comparison.rs [hunk 1]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+            src2_sign,
+            dst_bit,
+            subber,
         }
+    }
 
-        assert_eq!(trace.ret_events().len(), 1);
-        Prover::new(Box::new(GenericISA)).validate_witness(&trace)
+    fn as_any(&self) -> &dyn Any {
+        todo!()
```
Comment (sai-deng) [inline prover/src/opcodes/comparison.rs:667]: todo!() -> self

File: prover/src/opcodes/comparison.rs [hunk 2]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+    }
 
-        #[test]
-        fn test_sleiu_operations(
-            // Test both random values and specific edge cases
-            (src_val, imm_val) in prop_oneof![
-                // Random value pairs
-                (any::<u32>(), any::<u16>()),
+    fn as_any(&self) -> &dyn Any {
+        todo!()
```
Comment (sai-deng) [inline prover/src/opcodes/comparison.rs:871]: todo!() -> self

File: prover/src/opcodes/comparison.rs [hunk 3]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+        let src_val_packed = table.add_packed("src1_val_packed", src_val);
+
+        // Get the sign bits of src and imm
+        let src_sign = table.add_selected("src_sign", src_val, 31);
+        let imm_unpacked = state_cols.arg2_unpacked;
+        let imm_sign = table.add_selected("imm_sign", imm_unpacked, 15);
+
+        // We need to sign extend `imm`. First, get the sign bit and the necessary
+        // constants.
+        let imm_32b = table.add_zero_pad("imm_32b", imm_unpacked, 0);
```
Comment (sai-deng) [inline prover/src/opcodes/comparison.rs:1207]: same here

File: prover/src/opcodes/comparison.rs [hunk 4]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
-                Just((0, 1)),                  // Less than
-                Just((u32::MAX, u16::MAX)),    // Equal at max
-                Just((1, u16::MAX)),           // 1 < MAX
-                Just((0, u16::MAX)),           // Min < Max
-                Just((u32::MAX, 0)),           // Max > Min
-            ],
-        ) {
-            prop_assert!(test_imm_unsigned_comparisons_with_values(src_val, imm_val, "sltiu").is_ok());
+        // Get the sign bits of src and imm
+        let src_sign = table.add_selected("src_sign", src_val, 31);
```
Comment (sai-deng) [inline prover/src/opcodes/comparison.rs:797]: I think we can reuse `setup_sign_extended_immediate` defined in `integer_op.rs`

File: prover/src/opcodes/comparison.rs [hunk 5]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
-            // LDI events
-            (2, src1_val, 2),
-            (3, src2_val, 2),
-            // Initial values
-            (0, 0, 1),
-            (1, 0, 1),
-            // Comparison event
-            (4, expected, 1),
-        ];
+        // Direct comparison works whenever both sigs are equal. If not, it's determined
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:634]:
```suggestion
        // Direct comparison works whenever both signs are equal. If not, it's determined
```

File: prover/src/opcodes/comparison.rs [hunk 6]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
-                RET\n",
-            src1_val,
-            src2_val,
-            op.to_uppercase(),
+    fn name(&self) -> &'static str {
+        "SltTable"
+    }
+
+    // TODO: Consider swapping the order of src1 and src2 depending on the sign,
+    // or using a U32Add gadget.
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:586]: not sure I understand this comment, what would you like to achieve?

File: prover/src/opcodes/comparison.rs [hunk 7]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+            arg1: event.src1,
+            arg2: event.src2,
+        });
+        self.state_cols.populate(witness, state_rows)?;
+        self.subber.populate(witness)
+    }
+}
+/// SLTI table.
+///
+/// This table handles the SLT instruction, which performs signed
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:741]:
```suggestion
/// This table handles the SLTI instruction, which performs signed
```

File: prover/src/opcodes/comparison.rs [hunk 8]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+            expose_final_borrow: true, // we want the "underflow" bit out
+            commit_zout: false,        // we don't need the raw subtraction result
+        };
+        let subber = U32Sub::new(&mut table, src_val, signed_imm_32b, flags);
+        // `final_borrow` is 1 exactly when src_val < imm_val
+        let final_borrow: Col<B1> = subber
+            .final_borrow
+            .expect("Flag `expose_final_borrow` was set to `true`");
+
+        // Direct comparison works whenever both sigs are equal. If not, it's determined
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:837]:
```suggestion
        // Direct comparison works whenever both signs are equal. If not, it's determined
```

File: prover/src/opcodes/comparison.rs [hunk 9]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+    src2_sign: Col<B1>,
+    dst_bit: Col<B1>,
+    subber: U32Sub,
+}
+
+impl Table for SleTable {
+    type Event = SleEvent;
+
+    fn name(&self) -> &'static str {
+        "SltuTable"
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:993]:
```suggestion
        "SleTable"
```

File: prover/src/opcodes/comparison.rs [hunk 10]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+    }
+
+    fn new(cs: &mut ConstraintSystem, channels: &Channels) -> Self {
+        let mut table = cs.add_table("sltu");
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:997]:
```suggestion
        let mut table = cs.add_table("sle");
```

File: prover/src/opcodes/comparison.rs [hunk 11]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+    subber: U32Sub,
+}
 
-        assert_eq!(trace.ret_events().len(), 1);
-        Prover::new(Box::new(GenericISA)).validate_witness(&trace)
+impl Table for SltiTable {
+    type Event = SltiEvent;
+
+    fn name(&self) -> &'static str {
+        "SltTable"
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:764]:
```suggestion
        "SltiTable"
```

File: prover/src/opcodes/comparison.rs [hunk 12]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
     }
 
-    proptest! {
-        #![proptest_config(proptest::test_runner::Config::with_cases(20))]
+    // TODO: Consider swapping the order of src1 and src2 depending on the sign,
+    // or using a U32Add gadget.
+    fn new(cs: &mut ConstraintSystem, channels: &Channels) -> Self {
+        let mut table = cs.add_table("slt");
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:770]:
```suggestion
        let mut table = cs.add_table("slti");
```

File: prover/src/opcodes/comparison.rs [hunk 13]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+            expose_final_borrow: true, // we want the "underflow" bit out
+            commit_zout: false,        // we don't need the raw subtraction result
+        };
+        let subber = U32Sub::new(&mut table, src2_val, src1_val, flags);
+        // `final_borrow` is 1 exactly when src1_val < src2_val
+        let final_borrow: Col<B1> = subber
+            .final_borrow
+            .expect("Flag `expose_final_borrow` was set to `true`");
+
+        // Direct comparison works whenever both sigs are equal. If not, it's determined
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:1043]:
```suggestion
        // Direct comparison works whenever both signs are equal. If not, it's determined
```

File: prover/src/opcodes/comparison.rs [hunk 14]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+    ones: Col<B1, 32>,
+    dst_bit: Col<B1>,
+    subber: U32Sub,
+}
+
+impl Table for SleiTable {
+    type Event = SleiEvent;
+
+    fn name(&self) -> &'static str {
+        "SltuTable"
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:1170]:
```suggestion
        "SleiTable"
```

File: prover/src/opcodes/comparison.rs [hunk 15]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+    }
+
+    fn new(cs: &mut ConstraintSystem, channels: &Channels) -> Self {
+        let mut table = cs.add_table("sltu");
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:1174]:
```suggestion
        let mut table = cs.add_table("slei");
```

File: prover/src/opcodes/comparison.rs [hunk 16]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+    use anyhow::Result;
+    use petravm_asm::{isa::GenericISA, Opcode};
+    use proptest::prelude::*;
+    use proptest::prop_oneof;
+
+    use crate::prover::Prover;
+    use crate::test_utils::generate_trace;
+
+    /// Creates an execution trace for a simple program that uses either t
+    /// any comparison operation
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:1363]:
```suggestion
    /// Creates an execution trace for a simple program that uses a single
    /// comparison operation.
```

File: prover/src/opcodes/comparison.rs [hunk 17]
```diff
@@ -549,211 +557,1136 @@ impl TableFiller<ProverPackedField> for SleiuTable {
...
+                    "#[framesize(0x10)]\n\
+                 _start: 
+                    LDI.W @2, #{}\n\
+                    SLEI @3, @2, #{}\n\
+                    RET\n",
+                    src1_val, imm
+                )
+            }
+            _ => panic!("Not a comparison opcode"),
+        };
```
Comment (Nashtare) [inline prover/src/opcodes/comparison.rs:1451]: Would be a tad more succinct to do match only over opcodes first, as if you keep the distinction value / immediate these are the only diff in the `asm` variations. Would also remove all the casts `as u32` / `as u16` as currently done? But not blocking

File: prover/src/opcodes/comparison.rs (hunk 6 above)
Comment (4l0n50) [reply prover/src/opcodes/comparison.rs:586]: Old comment

File: prover/src/opcodes/comparison.rs (hunk 17 above)
Comment (4l0n50) [reply prover/src/opcodes/comparison.rs:1451]: What do you mean? Something like
```rust
        let asm_code = format!(
                "#[framesize(0x10)]\n\
                 _start: 
                    LDI.W @2, #{}\n\
                    {}
                    RET\n",
                match opcode {
                       Opcode::Sltu => ...
                }
            ),
```
?

Comment (sai-deng) [reply prover/src/opcodes/comparison.rs]: We can combine all these tests and combine instructions into one asm:

                 _start: 
                    LDI.W @2, #{}\n\
                    LDI.W @3, #{}\n\
                    SLTU @4, @2, @3\n\
                    SLTIU @5, @2, #{}\n\
...
                    RET\n",
                    
I can work on this. (I also have some fixes for witness filling for signed comparisons.)

**Review by sai-deng — APPROVED** [review]
